pyquil/api/_qpu.py

```diff
@@ -154,3 +182,63 @@ def _build_patch_values(self) -> dict:
             patch_table[k.name][k.index] = v
 
         return patch_table
+
+    def _build_arithmetic_patch_values(self) -> str:
+        """
+        Update the variables shim with the final values to be patched into the gate parameters,
+        according to the arithmetic expressions in the original program.
+
+        For example:
+
+            DECLARE theta REAL
+            RZ(3 * theta) 0
+
+        gets translated to:
+
+            DECLARE theta REAL
+            DECLARE __P REAL[0]
+            RZ(__P[0]) 0
+
+        and the recalculation table gets the entry: __P[0] = '3 * theta'. In our example, we
+        look up the value of 'theta' in the variables shim, multiply that by 3, and then add
+        an entry for __P[0] to the variables shim.
+
+        :return: Name of the region used to patch the parameters.
+        """
+        try:
+            recalculation_table = self._executable.recalculation_table
+        except AttributeError:
+            # No recalculation table, no work to be done here.
+            return ""
+
+        # We only declare one region to hold these values, and we need to add that name to
+        memory_reference_name = None
+        for memory_reference, expression in recalculation_table.items():
+            # Replace the memory references with any values the user has written
+            value = self._resolve_memory_references(expression)
+            self._variables_shim[memory_reference] = value
+            memory_reference_name = memory_reference.name
+        return memory_reference_name
+
+    def _resolve_memory_references(self, expression: Expression) -> Union[float, int]:
+        """
+        Traverse the given Expression, and replace any Memory References with whatever values
+        have been so far provided by the user for those memory spaces. Declared memory defaults
+        to zero.
+
+        :param expression: an Expression
+        """
+        if isinstance(expression, BinaryExp):
+            left = self._resolve_memory_references(expression.op1)
+            right = self._resolve_memory_references(expression.op2)
+            return expression.fn(left, right)
+        elif isinstance(expression, Function):
+            return expression.fn(self._resolve_memory_references(expression.expression))
+        elif isinstance(expression, Parameter):
+            raise ValueError(f"Unexpected Parameter in gate expression: {expression}")
+        elif isinstance(expression, float) or isinstance(expression, int):
```

Comment:
do we have to put all the numpy types here

Reply:
I don't know, how would we find out?

Reply:
try passing a numpy number?

```python
thetas = np.linspace(0,1,50)
qc.run(executable, memory_map={'thetas': thetas}
```